### ISSUE TITLE
[BUG] Using custom route with `datepublish`

### ISSUE THREAD

**Issue by dantleech** [bolt/bolt#5466]
When using the following route:

```
postbinding:
    path:           /{datepublish}/{slug}
    defaults:       { _controller: 'Bolt\Controllers\Frontend::record', 'contenttypeslug': 'post' }
    contenttype:    posts 
    requirements:
        datepublish:    '\d{4}-\d{2}-\d{2}'
```

The frontend renders fine, however when editing the content:

```
An exception has been thrown during the rendering of a template ("Some mandatory parameters are missing ("datepublish") to generate a URL for route "postbinding".") in "@bolt/editcontent/_buttons.twig" at line 60.
```

Details
-------

 - Relevant Bolt Version: 3.0
 - Install type: Composer install
 - PHP version: 7.0
 - Used webserver: php

Reproduction
------------

1. Using the default Bolt create an Entry.
2. Add the following route:

```yaml
postbinding:
    path:           /{datepublish}/{slug}
    defaults:       { _controller: 'Bolt\Controllers\Frontend::record', 'contenttypeslug': 'entry' }
    contenttype:    entries
    requirements:
        datepublish:    '\d{4}-\d{2}-\d{2}'
```

2. Edit one of the entries.

**Comment by GawainLynch** [comment]
Just checking (really tired eyes) but do you mean Bolt 2 as that controller namespace doesn't exists in v3, and the `routing.yml` format is from v2 as well.

**Comment by dantleech** [comment]
No, bolt 3. Maybe the docs are outdated? https://docs.bolt.cm/3.0/templates/templates-routes#contenttype-overrides

It does work on the frontend however

**Comment by GawainLynch** [comment]
> Maybe the docs are outdated

They are! Thank you :+1:

**Comment by dantleech** [comment]
If you can point me to a correct example I would be happy to make a PR.

**Comment by GawainLynch** [comment]
The examples in the [`routing.yml.dist`](https://github.com/bolt/bolt/blob/release/3.0/app/config/routing.yml.dist#L63-L69) should all be valid.

**Comment by dantleech** [comment]
hmm, what is the difference between those examples and mine?

**Comment by GawainLynch** [comment]
```yaml
    defaults:       { _controller: 'Bolt\Controllers\Frontend::record', 'contenttypeslug': 'post' }
```
versus…
```yaml
    defaults:
        _controller: controller.frontend:record
        contenttypeslug: page
```

The first is a deprecated format that is potentially problematic, and also refers to a plural namespace — `\Bolt\Controllers` instead of `\Bolt\Controller`

**Comment by dantleech** [comment]
Get the same error when using:

```
postbinding:
    path:           /{datepublish}/{slug}
    defaults:       
        _controller: controller.frontend:record
        contenttypeslug: 'post'
    contenttype: posts
    requirements:
        datepublish:    '\d{4}-\d{2}-\d{2}'
```

**Comment by SahAssar** [comment]
I got interested and went spelunking for this... the docs were added in [Aug 2013](https://github.com/bolt/docs/commit/0f12f66b1c56e5ebf68f5d3e84b50a844cb96efb), but the functionality doesn't seem to have been added until [Dec 2013](https://github.com/bolt/bolt/commit/8ec4248bc5a4e5e87ae7cd85461a1557023bbc00) and not much seems to have changed since then... So basically the problem was that the new storage tried to access the values via `$this->values[]` and that's why it worked on the frontend and not on the backend...

Anyway, this should get fixed by #5593, I just think it's fun when we have docs/functionality that is basically unchanged for three years :D
